src/main/java/org/candlepin/resource/util/EntitlementEnvironmentFilter.java

```diff
@@ -92,94 +53,128 @@ public EntitlementEnvironmentFilter setConsumerToBeUpdated(List<String> consumer
-                        contentProvideByLowerEnvs.addAll(mapOfEnvironmentContentUUID.get(env));
-                    }
+                        Set<String> contentOfLowerEnvs = contentOf(mapOfEnvironmentContentUUID, currentLower);
 
-                    for (String content : contentProvided) {
-                        if (entitlementContentUUIDs.contains(content) &&
-                            contentProvideByLowerEnvs.contains(content)) {
-                            entToRegen = true;
+                        entToRegen = contentProvided.stream()
+                            .filter(entitlementContentUUIDs::contains)
+                            .anyMatch(contentOfLowerEnvs::contains);
+
+                        if (entToRegen) {
                             break;
                         }
                     }
-
-                    if (entToRegen) {
-                        break;
-                    }
                 }
-            }
 
-            if (entToRegen) {
-                entitlementsToRegen.add(entitlementId);
+                if (entToRegen) {
+                    entitlementsToRegen.add(entitlementId);
+                }
             }
         }
 
         return entitlementsToRegen;
     }
 
+    private Set<String> contentOf(Map<String, List<String>> envContents, List<String> envIds) {
+        return envIds.stream()
+            .map(envContents::get)
+            .flatMap(Collection::stream)
+            .collect(Collectors.toSet());
+    }
+
+    private List<String> findCurrentLower(List<String> envRemoved, List<String> preExistingEnvs,
+        String environmentId) {
+        int fromIndex = envRemoved.indexOf(environmentId) + 1;
+        List<String> currentLower = new ArrayList<>(
+            preExistingEnvs.subList(fromIndex, preExistingEnvs.size()));
+        return (List<String>) CollectionUtils.subtract(currentLower, envRemoved);
+    }
+
+    private List<String> findCurrentHigher(List<String> updatedEnvs, String environmentId) {
+        if (updatedEnvs.contains(environmentId)) {
+            return updatedEnvs.subList(0, updatedEnvs.indexOf(environmentId));
+        }
+        else {
+            return updatedEnvs;
+        }
+    }
+
+    /**
+     * Returns the list of environment IDs which got added, removed or re-prioritized
+     * from consumers current existing environments.
+     *
+     * @return
+     *  Returns the unique environmentIds which got added, removed or re-prioritized
+     */
+    private Set<String> getEnvironmentChangeList(List<String> envRemoved,
+        List<String> preExistingEnvs, List<String> updatedEnvs) {
+
+        Set<String> environmentChangeList = new HashSet<>();
+
+        // Envs Added
+        environmentChangeList.addAll(CollectionUtils.subtract(updatedEnvs, preExistingEnvs));
+
+        // Env removed are kept in separate list for special use cases
+        envRemoved.addAll(CollectionUtils.subtract(preExistingEnvs, updatedEnvs));
```

Comment:
looking at the calling code, this might actually be a problem, because the envRemoved variable is shared between all consumers, so this `addAll` keeps adding environments of each consumer we loop over into it, and we don't want that (we want envRemoved to be per-consumer only)

Reply:
Good catch. Done.